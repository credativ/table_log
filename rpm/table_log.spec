# rpmbuild -D'pgmajorversion 14' -D'pginstdir /usr/pgsql-14' -ba rpm/table_log.spec

%global debug_package %{nil}
%global sname table_log

%if 0%{?rhel} && 0%{?rhel} == 7
%ifarch ppc64 ppc64le
%pgdg_set_ppc64le_compiler_at10
%endif
%endif

Summary:	PostgreSQL table log extension
Name:		%{sname}_%{pgmajorversion}
Version:	0.6.3
Release:	1%{?dist}
License:	BSD
Source0:	https://github.com/df7cb/table_log/archive/v%{version}.tar.gz
URL:		https://github.com/df7cb/table_log/
BuildRequires:	postgresql%{pgmajorversion}-devel pgdg-srpm-macros
Requires:	postgresql%{pgmajorversion}-server

%if 0%{?rhel} && 0%{?rhel} == 7
%ifarch ppc64 ppc64le
%pgdg_set_ppc64le_min_requires
%endif
%endif

%description
table_log is a PostgreSQL extension with a set of functions to log changes on
a table and to restore the state of the table or a specific row on any time in
the past.

%prep
%setup -q -n table_log-%{version}

%build
%if 0%{?rhel} && 0%{?rhel} == 7
%ifarch ppc64 ppc64le
	%pgdg_set_ppc64le_compiler_flags
%endif
%endif

USE_PGXS=1 PATH=%{pginstdir}/bin/:$PATH %{__make} %{?_smp_mflags}

%install
%{__rm} -rf %{buildroot}
USE_PGXS=1 PATH=%{pginstdir}/bin/:$PATH %{__make} DESTDIR=%{buildroot} %{?_smp_mflags} install
%{__install} -d %{buildroot}%{pginstdir}/share/extension
%{__install} -d %{buildroot}%{pginstdir}/bin

%clean
%{__rm} -rf %{buildroot}

%files
%defattr(644,root,root,755)
%doc %{pginstdir}/doc/extension/table_log.md
%{pginstdir}/lib/table_log*
%{pginstdir}/lib/bitcode/table_log*
%{pginstdir}/share/extension/table_log*.sql*
%{pginstdir}/share/extension/table_log.control

%changelog
* Thu Jan 20 2022 Christoph Berg <myon@debian.org> - 0.6.3-1
- Initial packaging for PostgreSQL RPM Repository

