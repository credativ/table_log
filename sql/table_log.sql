CREATE EXTENSION table_log;
SET client_min_messages TO warning;

-- drop old trigger
DROP TRIGGER test_log_chg ON test; -- ignore any error

-- create demo table
DROP TABLE test; -- ignore any error
CREATE TABLE test (
  id                    INT                 NOT NULL
                                            PRIMARY KEY,
  name                  VARCHAR(20)         NOT NULL
);

-- create the table without data from demo table
DROP TABLE test_log; -- ignore any error
SELECT * INTO test_log FROM test LIMIT 0;
ALTER TABLE test_log ADD COLUMN trigger_mode VARCHAR(10);
ALTER TABLE test_log ADD COLUMN trigger_tuple VARCHAR(5);
ALTER TABLE test_log ADD COLUMN trigger_changed TIMESTAMPTZ;
ALTER TABLE test_log ADD COLUMN trigger_id BIGINT;
CREATE SEQUENCE test_log_id;
SELECT SETVAL('test_log_id', 1, FALSE);
ALTER TABLE test_log ALTER COLUMN trigger_id SET DEFAULT NEXTVAL('test_log_id');

-- create trigger
CREATE TRIGGER test_log_chg AFTER UPDATE OR INSERT OR DELETE ON test FOR EACH ROW
               EXECUTE PROCEDURE table_log();

-- test trigger
INSERT INTO test VALUES (1, 'name');
SELECT id, name FROM test;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
UPDATE test SET name='other name' WHERE id=1;
SELECT id, name FROM test;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;

-- create restore table
SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW());
SELECT id, name FROM test_recover;

DROP TABLE test;
DROP TABLE test_log;
DROP TABLE test_recover;

-- test table_log_init with all arguments
-- trigger_user and trigger_changed might differ, so ignore it

SET client_min_messages TO warning;

CREATE TABLE test(id integer, name text);
SELECT table_log_init(5, 'test');
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
DROP TABLE test;
DROP TABLE test_log;
DROP SEQUENCE test_log_seq;

CREATE TABLE test(id integer, name text);
SELECT table_log_init(4, 'test');
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
DROP TABLE test;
DROP TABLE test_log;
DROP SEQUENCE test_log_seq;

CREATE TABLE test(id integer, name text);
SELECT table_log_init(3, 'test');
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple FROM test_log;
DROP TABLE test;
DROP TABLE test_log;

-- Check table_log_restore_table()

CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'test');
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;

SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW(), '2', NULL::int, 1);
SELECT id, name FROM test_recover;

DROP TABLE test;
DROP TABLE test_log;
DROP TABLE test_recover;
DROP SEQUENCE test_log_seq;

-- Check partition support with auto-generated
-- log table name.
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'public', NULL, 'PARTITION');

SET table_log.active_partition = 0;

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');

SET table_log.active_partition = 1;

UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;

SELECT id, name FROM test_log;
SELECT id, name FROM test_log_0;
SELECT id, name FROM test_log_1;

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW(), '3', NULL::int, 1);
SELECT id, name FROM test_recover;

DROP TABLE test;
DROP VIEW  test_log;
DROP TABLE test_log_0;
DROP TABLE test_log_1;
DROP TABLE test_recover;
DROP SEQUENCE test_log_seq;

--
-- Check with schema-qualified restore and log table
--
CREATE SCHEMA log;

CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION');

SET table_log.active_partition = 0;

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');

SET table_log.active_partition = 1;

UPDATE test SET name = 'veronica' WHERE id = 3;

SELECT * FROM test;

DELETE FROM test WHERE id = 1;

SELECT * FROM test;

SELECT id, name FROM log.test_log;
SELECT id, name FROM log.test_log_0;
SELECT id, name FROM log.test_log_1;

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'log.test_log', 'trigger_id', 'log.test_recover', NOW(), '3', NULL::int, 1);
SELECT id, name FROM log.test_recover;

DROP SCHEMA log CASCADE;
DROP TABLE test;

--
-- Test with a schema-qualified and case sensitive log and
-- restore table
--

CREATE SCHEMA log;

CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'log', 'Test', 'PARTITION');

-- Check for log partitions
SELECT 'log."Test_0"'::regclass;
SELECT 'log."Test_1"'::regclass;

SET table_log.active_partition = 0;

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');

SET table_log.active_partition = 1;

UPDATE test SET name = 'veronica' WHERE id = 3;

SELECT * FROM test;

DELETE FROM test WHERE id = 1;

SELECT * FROM test;

SELECT id, name FROM log."Test";
SELECT id, name FROM log."Test_0";
SELECT id, name FROM log."Test_1";

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'log."Test"', 'trigger_id', 'log."Test_recover"', NOW(), '3', NULL::int, 1);
SELECT id, name FROM log."Test_recover";

DROP SCHEMA log CASCADE;
DROP TABLE test;

--
-- Test basic log mode
--
CREATE SCHEMA log;

CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);

-- this should fail, no trigger actions specified
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION', true, '{}');

-- this should succeed, but leave out inserts
-- generate the log table this time...
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION', true, '{UPDATE, DELETE}');

-- Log partitions created?
SELECT 'log.test_log_0'::regclass;
SELECT 'log.test_log_1'::regclass;

SET table_log.active_partition = 0;

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');

SET table_log.active_partition = 1;

UPDATE test SET name = 'veronica' WHERE id = 3;

SELECT * FROM test;
DELETE FROM test WHERE id = 1;
SELECT * FROM test;

-- UPDATE logged only, but only old tuples
SELECT id, name FROM log.test_log;
SELECT id, name FROM log.test_log_0;
SELECT id, name FROM log.test_log_1;

-- NOTE:
--    We don't test table_log_restore_table() at this point, since
--    restore from data collected by table_log_basic() is not yet supported.

DROP SCHEMA log CASCADE;
DROP TABLE test;

RESET client_min_messages;

