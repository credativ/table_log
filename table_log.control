comment = 'Module to log changes on tables'
default_version = '0.6.1'
module_pathname = '$libdir/table_log'
relocatable = false
