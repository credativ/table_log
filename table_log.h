/*
 * table_log () -- log changes to another table
 *
 *
 * see README.table_log for details
 *
 *
 * written by Andreas ' ads' Scherbaum (ads@pgug.de)
 * adapted for PostgreSQL 9.1+ by Bernd Helmle (bernd.helmle@credativ.de)
 *
 */
#define MAX_TABLE_LOG_PARTITIONS 2

/*
 * Selected log table identifier, relies
 * on the current selected partition via table_log.active_partition
 */
typedef int TableLogPartitionId;
