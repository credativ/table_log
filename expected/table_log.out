CREATE EXTENSION table_log;
SET client_min_messages TO warning;
-- drop old trigger
DROP TRIGGER test_log_chg ON test; -- ignore any error
ERROR:  relation "test" does not exist
-- create demo table
DROP TABLE test; -- ignore any error
ERROR:  table "test" does not exist
CREATE TABLE test (
  id                    INT                 NOT NULL
                                            PRIMARY KEY,
  name                  VARCHAR(20)         NOT NULL
);
-- create the table without data from demo table
DROP TABLE test_log; -- ignore any error
ERROR:  table "test_log" does not exist
SELECT * INTO test_log FROM test LIMIT 0;
ALTER TABLE test_log ADD COLUMN trigger_mode VARCHAR(10);
ALTER TABLE test_log ADD COLUMN trigger_tuple VARCHAR(5);
ALTER TABLE test_log ADD COLUMN trigger_changed TIMESTAMPTZ;
ALTER TABLE test_log ADD COLUMN trigger_id BIGINT;
CREATE SEQUENCE test_log_id;
SELECT SETVAL('test_log_id', 1, FALSE);
 setval 
--------
      1
(1 row)

ALTER TABLE test_log ALTER COLUMN trigger_id SET DEFAULT NEXTVAL('test_log_id');
-- create trigger
CREATE TRIGGER test_log_chg AFTER UPDATE OR INSERT OR DELETE ON test FOR EACH ROW
               EXECUTE PROCEDURE table_log();
-- test trigger
INSERT INTO test VALUES (1, 'name');
SELECT id, name FROM test;
 id | name 
----+------
  1 | name
(1 row)

SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
 id | name | trigger_mode | trigger_tuple | trigger_id 
----+------+--------------+---------------+------------
  1 | name | INSERT       | new           |          1
(1 row)

UPDATE test SET name='other name' WHERE id=1;
SELECT id, name FROM test;
 id |    name    
----+------------
  1 | other name
(1 row)

SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
 id |    name    | trigger_mode | trigger_tuple | trigger_id 
----+------------+--------------+---------------+------------
  1 | name       | INSERT       | new           |          1
  1 | name       | UPDATE       | old           |          2
  1 | other name | UPDATE       | new           |          3
(3 rows)

-- create restore table
SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW());
 table_log_restore_table 
-------------------------
 test_recover
(1 row)

SELECT id, name FROM test_recover;
 id |    name    
----+------------
  1 | other name
(1 row)

DROP TABLE test;
DROP TABLE test_log;
DROP TABLE test_recover;
-- test table_log_init with all arguments
-- trigger_user and trigger_changed might differ, so ignore it
SET client_min_messages TO warning;
CREATE TABLE test(id integer, name text);
SELECT table_log_init(5, 'test');
 table_log_init 
----------------
 
(1 row)

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
 id |   name   | trigger_mode | trigger_tuple | trigger_id 
----+----------+--------------+---------------+------------
  1 | joe      | INSERT       | new           |          1
  2 | barney   | INSERT       | new           |          2
  3 | monica   | INSERT       | new           |          3
  3 | monica   | UPDATE       | old           |          4
  3 | veronica | UPDATE       | new           |          5
  1 | joe      | DELETE       | old           |          6
(6 rows)

DROP TABLE test;
DROP TABLE test_log;
DROP SEQUENCE test_log_seq;
CREATE TABLE test(id integer, name text);
SELECT table_log_init(4, 'test');
 table_log_init 
----------------
 
(1 row)

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple, trigger_id FROM test_log;
 id |   name   | trigger_mode | trigger_tuple | trigger_id 
----+----------+--------------+---------------+------------
  1 | joe      | INSERT       | new           |          1
  2 | barney   | INSERT       | new           |          2
  3 | monica   | INSERT       | new           |          3
  3 | monica   | UPDATE       | old           |          4
  3 | veronica | UPDATE       | new           |          5
  1 | joe      | DELETE       | old           |          6
(6 rows)

DROP TABLE test;
DROP TABLE test_log;
DROP SEQUENCE test_log_seq;
CREATE TABLE test(id integer, name text);
SELECT table_log_init(3, 'test');
 table_log_init 
----------------
 
(1 row)

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name, trigger_mode, trigger_tuple FROM test_log;
 id |   name   | trigger_mode | trigger_tuple 
----+----------+--------------+---------------
  1 | joe      | INSERT       | new
  2 | barney   | INSERT       | new
  3 | monica   | INSERT       | new
  3 | monica   | UPDATE       | old
  3 | veronica | UPDATE       | new
  1 | joe      | DELETE       | old
(6 rows)

DROP TABLE test;
DROP TABLE test_log;
-- Check table_log_restore_table()
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'test');
 table_log_init 
----------------
 
(1 row)

INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW(), '2', NULL::int, 1);
 table_log_restore_table 
-------------------------
 test_recover
(1 row)

SELECT id, name FROM test_recover;
 id |  name  
----+--------
  2 | barney
(1 row)

DROP TABLE test;
DROP TABLE test_log;
DROP TABLE test_recover;
DROP SEQUENCE test_log_seq;
-- Check partition support with auto-generated
-- log table name.
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'public', NULL, 'PARTITION');
 table_log_init 
----------------
 
(1 row)

SET table_log.active_partition = 0;
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
SET table_log.active_partition = 1;
UPDATE test SET name = 'veronica' WHERE id = 3;
DELETE FROM test WHERE id = 1;
SELECT id, name FROM test_log;
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | monica
  3 | monica
  3 | veronica
  1 | joe
(6 rows)

SELECT id, name FROM test_log_0;
 id |  name  
----+--------
  1 | joe
  2 | barney
  3 | monica
(3 rows)

SELECT id, name FROM test_log_1;
 id |   name   
----+----------
  3 | monica
  3 | veronica
  1 | joe
(3 rows)

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'test_log', 'trigger_id', 'test_recover', NOW(), '3', NULL::int, 1);
 table_log_restore_table 
-------------------------
 test_recover
(1 row)

SELECT id, name FROM test_recover;
 id |   name   
----+----------
  3 | veronica
(1 row)

DROP TABLE test;
DROP VIEW  test_log;
DROP TABLE test_log_0;
DROP TABLE test_log_1;
DROP TABLE test_recover;
DROP SEQUENCE test_log_seq;
--
-- Check with schema-qualified restore and log table
--
CREATE SCHEMA log;
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION');
 table_log_init 
----------------
 
(1 row)

SET table_log.active_partition = 0;
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
SET table_log.active_partition = 1;
UPDATE test SET name = 'veronica' WHERE id = 3;
SELECT * FROM test;
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | veronica
(3 rows)

DELETE FROM test WHERE id = 1;
SELECT * FROM test;
 id |   name   
----+----------
  2 | barney
  3 | veronica
(2 rows)

SELECT id, name FROM log.test_log;
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | monica
  3 | monica
  3 | veronica
  1 | joe
(6 rows)

SELECT id, name FROM log.test_log_0;
 id |  name  
----+--------
  1 | joe
  2 | barney
  3 | monica
(3 rows)

SELECT id, name FROM log.test_log_1;
 id |   name   
----+----------
  3 | monica
  3 | veronica
  1 | joe
(3 rows)

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'log.test_log', 'trigger_id', 'log.test_recover', NOW(), '3', NULL::int, 1);
 table_log_restore_table 
-------------------------
 log.test_recover
(1 row)

SELECT id, name FROM log.test_recover;
 id |   name   
----+----------
  3 | veronica
(1 row)

DROP SCHEMA log CASCADE;
DROP TABLE test;
--
-- Test with a schema-qualified and case sensitive log and
-- restore table
--
CREATE SCHEMA log;
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
SELECT table_log_init(5, 'public', 'test', 'log', 'Test', 'PARTITION');
 table_log_init 
----------------
 
(1 row)

-- Check for log partitions
SELECT 'log."Test_0"'::regclass;
   regclass   
--------------
 log."Test_0"
(1 row)

SELECT 'log."Test_1"'::regclass;
   regclass   
--------------
 log."Test_1"
(1 row)

SET table_log.active_partition = 0;
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
SET table_log.active_partition = 1;
UPDATE test SET name = 'veronica' WHERE id = 3;
SELECT * FROM test;
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | veronica
(3 rows)

DELETE FROM test WHERE id = 1;
SELECT * FROM test;
 id |   name   
----+----------
  2 | barney
  3 | veronica
(2 rows)

SELECT id, name FROM log."Test";
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | monica
  3 | monica
  3 | veronica
  1 | joe
(6 rows)

SELECT id, name FROM log."Test_0";
 id |  name  
----+--------
  1 | joe
  2 | barney
  3 | monica
(3 rows)

SELECT id, name FROM log."Test_1";
 id |   name   
----+----------
  3 | monica
  3 | veronica
  1 | joe
(3 rows)

--
-- NOTE: since we use partitioning here, we use the special log
--       view to restore the data with id = 2
--
SELECT table_log_restore_table('test', 'id', 'log."Test"', 'trigger_id', 'log."Test_recover"', NOW(), '3', NULL::int, 1);
 table_log_restore_table 
-------------------------
 log."Test_recover"
(1 row)

SELECT id, name FROM log."Test_recover";
 id |   name   
----+----------
  3 | veronica
(1 row)

DROP SCHEMA log CASCADE;
DROP TABLE test;
--
-- Test basic log mode
--
CREATE SCHEMA log;
CREATE TABLE test(id integer, name text);
ALTER TABLE test ADD PRIMARY KEY(id);
-- this should fail, no trigger actions specified
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION', true, '{}');
ERROR:  table_log_init: at least one trigger action must be specified
-- this should succeed, but leave out inserts
-- generate the log table this time...
SELECT table_log_init(5, 'public', 'test', 'log', NULL, 'PARTITION', true, '{UPDATE, DELETE}');
 table_log_init 
----------------
 
(1 row)

-- Log partitions created?
SELECT 'log.test_log_0'::regclass;
    regclass    
----------------
 log.test_log_0
(1 row)

SELECT 'log.test_log_1'::regclass;
    regclass    
----------------
 log.test_log_1
(1 row)

SET table_log.active_partition = 0;
INSERT INTO test VALUES(1, 'joe');
INSERT INTO test VALUES(2, 'barney');
INSERT INTO test VALUES(3, 'monica');
SET table_log.active_partition = 1;
UPDATE test SET name = 'veronica' WHERE id = 3;
SELECT * FROM test;
 id |   name   
----+----------
  1 | joe
  2 | barney
  3 | veronica
(3 rows)

DELETE FROM test WHERE id = 1;
SELECT * FROM test;
 id |   name   
----+----------
  2 | barney
  3 | veronica
(2 rows)

-- UPDATE logged only, but only old tuples
SELECT id, name FROM log.test_log;
 id |  name  
----+--------
  3 | monica
  1 | joe
(2 rows)

SELECT id, name FROM log.test_log_0;
 id | name 
----+------
(0 rows)

SELECT id, name FROM log.test_log_1;
 id |  name  
----+--------
  3 | monica
  1 | joe
(2 rows)

-- NOTE:
--    We don't test table_log_restore_table() at this point, since
--    restore from data collected by table_log_basic() is not yet supported.
DROP SCHEMA log CASCADE;
DROP TABLE test;
RESET client_min_messages;
